volatility3/framework/plugins/windows/amcache.py

```diff
@@ -0,0 +1,603 @@
+# This file is Copyright 2019 Volatility Foundation and licensed under the Volatility Software License 1.0
+# which is available at https://www.volatilityfoundation.org/license/vsl-v1.0
+#
+
+import datetime
+import enum
+import logging
+from itertools import groupby
+from typing import Any, Dict, Iterator, List, NamedTuple, Optional, Tuple, Union
+
+from volatility3.framework import exceptions, interfaces, renderers
+from volatility3.framework.configuration import requirements
+from volatility3.framework.layers import registry
+from volatility3.framework.renderers import conversion
+from volatility3.framework.symbols.windows.extensions import registry as reg_extensions
+from volatility3.plugins import timeliner
+from volatility3.plugins.windows.registry import hivelist
+
+vollog = logging.getLogger(__name__)
+
+
+class AmcacheEntryType(enum.IntEnum):
+    Driver = 1
+    Program = 2
+    File = 3
+
+
+AmcacheEntry = NamedTuple(
+    "AmcacheEntry",
+    [
+        ("Path", Union[str, interfaces.renderers.BaseAbsentValue]),
+        ("Company", Union[str, interfaces.renderers.BaseAbsentValue]),
+        (
+            "LastModifyTime",
+            Union[datetime.datetime, interfaces.renderers.BaseAbsentValue],
+        ),
+        (
+            "LastModifyTime2",
+            Union[datetime.datetime, interfaces.renderers.BaseAbsentValue],
+        ),
+        ("InstallTime", Union[datetime.datetime, interfaces.renderers.BaseAbsentValue]),
+        ("CompileTime", Union[datetime.datetime, interfaces.renderers.BaseAbsentValue]),
+        ("SHA1", Union[str, interfaces.renderers.BaseAbsentValue]),
+        ("Service", Union[str, interfaces.renderers.BaseAbsentValue]),
+        ("ProductName", Union[str, interfaces.renderers.BaseAbsentValue]),
+        ("EntryType", str),
+    ],
+)
+
+
+def _try_get_value(
+    values: List[reg_extensions.CM_KEY_VALUE],
+    name: str,
+    expected_type: reg_extensions.RegValueTypes,
+) -> Any:
+    try:
+        value = next(
```

Comment:
So I can't quite see why this is using next rather than just a 0 index?  It looks like you're just using the head of the list because every path from this next returns (meaning you never get to the second item)?  I don't specifically have anything against it, but it makes me wonder what's going on (and why it's provided a list of values) if it only ever uses the first one?  I'd have gone with a list comprehension and a 0 index (if the list has any contents)...

Reply:
The advantage of `next` with a generator comprehension over using a list comprehension with a zero-index is that creating a complete list of values using a list comprehension requires performing the `.get_name().lower().strip().strip("\u0000")` operations on every value, even those in the iterator beyond the one we're looking for. Using `next` with a generator expression only performs those operations as needed, and stops evaluating once the desired value is found.

Reply:
You're right about the efficiency.  I still find standalone calls to `next` a little unusual (rather than a `for...: break`) but that's ok, it just threw me a little (and made me wonder why we stopped after the first one, what are the others used for, but a for loop would have the same effect))?